File: node/sample/ab/ab.js
var etsm = require('etsm');

class Foo {
    constructor() {
        this.sm = new etsm.StateMachine(this);
        this.a = new etsm.State(this.EnterA, this.ExitA);
        this.b = new etsm.State(this.EnterB, null);
    }

    EnterA(){
        process.stdout.write(' ->A ');
    }

    ExitA(){
        process.stdout.write(' A-> ');
    }

    EnterB(){
        process.stdout.write(' ->B ');
    }

    // don't implement Exit for B state
    // ExitB(){
    //     process.stdout.write(' B-> ');
    //     assert(this.sm.Transition(this.a) == false);
    // }

    Test() {
        this.sm.Transition(this.a);
        this.sm.Transition(this.b);
        this.sm.Transition(null);
    }    
}

foo = new Foo();
foo.Test();

File: lua/sample/ab/ab.lua
package.path = './../../etsm/?.lua;' .. package.path
local etsm = require 'etsm'

Foo = {}

function Foo:new(o)
    self.__index = self
    return setmetatable({
        sm = etsm.StateMachine:new(self),
        a = etsm.State:new(Foo.EnterA, Foo.ExitA),
        b = etsm.State:new(Foo.EnterB, nil)
    }, self)
end

function Foo:EnterA()
    io.write(" ->A ")
end

function Foo:ExitA()
    io.write(" A-> ")
end

function Foo:EnterB()
    io.write(" ->B ")
end

-- no exit for state B
-- function Foo:ExitB()
--     io.write(" B-> ")
-- end

function Foo:Test()
    self.sm:Transition(self.a)
    self.sm:Transition(self.b)
    self.sm:Transition(nil)
end

local foo = Foo:new(nil)
foo:Test()
